pages/packingDetailed.js

```diff
@@ -0,0 +1,333 @@
+  const f = (u) =>
+    fetch(u, {
+      headers: { "Content-Type": "application/json", Authorization: token },
+    }).then((r) => r.json());
+
+  if (urls.length > 1) {
+    return Promise.all(urls.map(f));
+  }
+  return f(urls);
+}
+
+class PackingOrder extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      orderId: props.data[0].orderId,
+      delivery_date: props.data[0].delivery_date,
+      pantryNote: props.data[0].pantry_note,
+      guestNote: props.data[0].guestNote,
+      firstName: props.data[0].firstName,
+      lastInitial: props.data[0].lastInitial,
+      status: props.data[0].status,
+      order_timestamp: props.data[0].order_timestamp,
+      items: props.data[0].items,
+      itemMap: props.data[1],
+      error: null,
+      success: null,
+    };
+  }
+
+  savePantryNote = (newPantryNote = this.state.pantryNote) => {
+    this.state.pantryNote = newPantryNote;
+    this.setState({ pantryNote: this.state.pantryNote });
+    fetch("/api/orders/SetPantryNote", {
+      method: "POST",
+      body: JSON.stringify({
+        orderId: this.state.orderId,
+        message: this.state.pantryNote,
+      }),
+      headers: { "Content-Type": "application/json", Authorization: token },
+    }).then(() => {
+      this.setState({ success: "Saved pantry note successfully!" });
+      setTimeout(() => this.setState({ success: null }), 1000);
+    });
+  };
+
+  cancelPantryNote = () => {
+    document.getElementById("pantry_note").value = this.state.pantryNote;
+  };
+
+  changeOrderStatus = () => {
+    var newStatus = "";
+    if (this.state.status == ORDER_STATUS_OPEN) {
+      newStatus = ORDER_STATUS_PROCESSING;
+    } else if (this.state.status == ORDER_STATUS_PROCESSING) {
+      newStatus = ORDER_STATUS_COMPLETE;
+    } else {
+      newStatus = ORDER_STATUS_PROCESSING;
+    }
+    this.state.status = newStatus;
+    this.setState({ status: this.state.status });
+    fetch("/api/orders/SetOrderStatus", {
+      method: "POST",
+      body: JSON.stringify({
+        orderId: this.state.orderId,
+        status: this.state.status,
+      }),
+      headers: { "Content-Type": "application/json", Authorization: token },
+    }).then(() => {
+      this.setState({ success: "Changed order status successfully!" });
+      setTimeout(() => this.setState({ success: null }), 1000);
+    });
+  };
+
+  changeStatusOfItem(barcode) {
+    this.state.items[barcode].isPacked = !this.state.items[barcode].isPacked;
+    this.setState({ items: this.state.items });
+    fetch("/api/orders/SetOrderItemStatus", {
+      method: "POST",
+      body: JSON.stringify({
+        orderId: this.state.orderId,
+        itemId: barcode,
+        isPacked: this.state.items[barcode].isPacked,
+      }),
+      headers: { "Content-Type": "application/json", Authorization: token },
+    }).then(() => {});
+  }
+
+  displayUnpackedItemRow(barcode, value) {
+    var itemName = this.state.itemMap[barcode].itemName;
```

Comment:
We can make this more resilient to failure by doing `this.state.itemMap[barcode]?.itemName`. This way, if `this.state.itemMap[barcode]` doesn't exist the page won't crash (accessing `.itemName` of `undefined` is an error), it'll just set `itemName` to null.